### ISSUE TITLE
What to do to optimize CI?

### ISSUE THREAD

**Issue by mamcx** [cachix/devenv#1575]
I'm on Circleci and hope that moving to `devenv` improves my CI but takes longer instead.

Of course, I don't know exactly how to speed up the process.

So, is there a guideline on how to speed up this? What to cache? 

Now I have:

```yaml
commands:
  restore_cache_devenv:
    steps:
      - restore_cache:
          key: cache_devenv-v1-{{ arch }}-{{ checksum "devenv.lock" }}

  save_cache_devenv:
    steps:
      - save_cache:
          key: cache_devenv-v1-{{ arch }}-{{ checksum "devenv.lock" }}
          paths:
            - "./.devenv"

  setup_nix:
    description: "Setup nix environment"
    steps:
      - checkout
      - restore_cache_devenv
      - run:
          name: Install devenv
          command: |
            nix-env -iA devenv -f https://github.com/NixOS/nixpkgs/tarball/nixpkgs-unstable
            devenv info
      - save_cache_devenv
```

**Comment by domenkozar** [comment]
@mamcx I assume most of the time you'd like to save is in `setup_nix`?

**Comment by byrnedo** [comment]
Also interested in this. 
Does caching the `.devenv` dirs help? I've been toying with it in github actions but getting some file not found errors.

**Comment by extrange** [comment]
I'm getting the same issue - devenv takes 7-8 minutes to setup the tests (running tests is fast). For reference my workflow file is

```yaml
name: Run devenv test
on:
  push:
  workflow_call:
    inputs:
      cache-size:
        required: false
        type: string
        default: 3G

jobs:
  tests:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4

      - uses: nixbuild/nix-quick-install-action@v31
        with:
          nix_conf: |
            keep-env-derivations = true
            keep-outputs = true

      - name: Restore and save Nix store
        uses: nix-community/cache-nix-action@v6
        with:
          primary-key: nix-${{ runner.os }}-${{ hashFiles('**/*.nix', '**/devenv.lock') }}
          restore-prefixes-first-match: nix-${{ runner.os }}-
          gc-max-store-size-linux: ${{ inputs.cache-size }}

      - name: Install devenv.sh
        run: nix profile install nixpkgs#devenv

      - name: Test # Most of the delay is here
        run: devenv test
```

**Comment by leiserfg** [comment]
That's too long, most probably you are building packages that are not cached. To me it takes like one or two mins, not that bad but not good either.

**Comment by davidnbr** [comment]
You could make sure that you are using `cachix.pull = [ "cache-name" ]` or the extra substituters are correctly set for the built packages.

**Comment by zevisert** [comment]
TL;DR Caching in github actions involves downloading assets, which is what is happening when pulling from cache.nixos.org, so would putting that cache in another place really actually save any time in the end?

--- 

Broadly, this is due to github actions runners having a bunch of non-nix pre-installed tools, but devenv has to pull all of the nix-built versions for itself, right? Like just installing devenv (without setting up my environment) takes 40 seconds because of all of these downloads: 

```log
$ nix profile add nixpkgs#devenv
unpacking 'github:NixOS/nixpkgs/01fbdeef22b76df85ea168fbfe1bfd9e63681b30' into the Git cache...
these 107 paths will be fetched (168.0 MiB download, 865.0 MiB unpacked):
  /nix/store/bifwnk7bvspj2mqc1w7dcydz8lkvfy1q-acl-2.3.2
  /nix/store/sl325r3s6d4gl9lmr2hwg2wgp2m0qs7c-attr-2.5.2
  ...
  /nix/store/ixhlv41i2wpl84xgjcks061dz4yssbg3-zlib-1.3.2
  /nix/store/k0rqiflg1vkn1kj96br5pfxj40p3srz4-zstd-1.5.7
```

Then after that, I see devenv gets to work setting up my environment (for 238s) to run my git hooks which only take <1s:

```log
$ devenv tasks run devenv:git-hooks:run
Configuring shell
Configuring shell in 238s
Loading tasks
Loading tasks in 2.24ms
Running tasks     devenv:enterTest
Running tasks
Running           devenv:files:cleanup
Succeeded         devenv:files:cleanup (4.29ms)
Running           devenv:files
Succeeded         devenv:files (18.26ms)
Running           devenv:git-hooks:install
Succeeded         devenv:git-hooks:install (15.96ms)
Running           devenv:git-hooks:run
Running           devenv:enterShell
Succeeded         devenv:enterShell (16.27ms)
Succeeded         devenv:git-hooks:run (536.87ms)
Running tasks in 578ms
5 Succeeded
```

Most of the time was spent by devenv "configuring the shell" - and maybe there's a way to cache that better in github actions. But for the most part, the work in configuring the shell is just more downloads from cache.nixos.org, right? Would pulling from a github actions cache be any different than pulling from that? Would it maybe be slower, since the nix packages have to be uploaded to github actions cache at the end?

**Comment by sandydoo** [comment]
There's no magic solution here, at least not with the traditional CI providers. Nix places additional resource load on the often underpowered CI machines. Things you can do in order of complexity:

- Add a personal cache. [cachix-action](https://github.com/cachix/cachix-action/) or the cachix devenv integration with a free tier from [cachix.org](https://cachix.org/).
- Use more powerful runners. Bigger runners = more CPU and crucially disk I/O.
- Cache `/nix/store`, flake fetcher caches, etc. Can be a pain to maintain. It's _crucial_ that the cache is fast, otherwise this will make things even worse (see next section).
- Self-hosted runner. Pre-install Nix and keep a hot Nix store.

> Would pulling from a Github Actions cache be faster than downloading from cache.nixos.org?

Avoid tools and actions that use the GitHub Actions cache. It's extremely slow and aggressively rate-limited. At the very least test the before and after instead of blindly adding these to your workflow.

> Should I cache `.devenv`?

I would advise against it. Even though it contains an evaluation cache that may make sense to cache in between runs, the operational complexity is not worth it IMO.

There's a lot of work happening in this space, including on our end. We'll announce better approaches as they become available.
